### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'encoder'

### ISSUE THREAD

**Issue by KhinChoWin-AI** [altndrr/clup#4]
Clustering 15339 samples
num_samples_class=4601
num_samples_class=0
num_samples_class=0
num_samples_class=0
num_samples_class=0
num_samples_class=0
num_samples_class=0
Selected a total of 4601 samples
Using 16bit native Automatic Mixed Precision (AMP)
Trainer already configured with model summary callbacks: [<class 'pytorch_lightning.callbacks.rich_model_summary.RichModelSummary'>]. Skipping setting a default `ModelSummary` callback.
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
before init in systems.clup
model:  None
╭───────────────────────────── Traceback (most recent call last) ──────────────────────────────╮
│                                                                                              │
│ /home1/ai20resch16001/anaconda3/lib/python3.9/runpy.py:197 in _run_module_as_main            │
│                                                                                              │
│   194 │   main_globals = sys.modules["__main__"].__dict__                                    │
│   195 │   if alter_argv:                                                                     │
│   196 │   │   sys.argv[0] = mod_spec.origin                                                  │
│ ❱ 197 │   return _run_code(code, main_globals, None,                                         │
│   198 │   │   │   │   │    "__main__", mod_spec)                                             │
│   199                                                                                        │
│   200 def run_module(mod_name, init_globals=None,                                            │
│ /home1/ai20resch16001/anaconda3/lib/python3.9/runpy.py:87 in _run_code                       │
│                                                                                              │
│    84 │   │   │   │   │      __loader__ = loader,                                            │
│    85 │   │   │   │   │      __package__ = pkg_name,                                         │
│    86 │   │   │   │   │      __spec__ = mod_spec)                                            │
│ ❱  87 │   exec(code, run_globals)                                                            │
│    88 │   return run_globals                                                                 │
│    89                                                                                        │
│    90 def _run_module_code(code, init_globals=None,                                          │
│                                                                                              │
│ /home1/ai20resch16001/clup/cli/__main__.py:54 in <module>                                    │
│                                                                                              │
│   51 │   │   │   command = COMMANDS.get(key)                                                 │
│   52 │   │   │   if command and value:                                                       │
│   53 │   │   │   │   cmd = command(options, console)                                         │
│ ❱ 54 │   │   │   │   cmd.run()                                                               │
│   55 │   │   │   │   print("after run")                                                      │
│   56 │   │   │   │   executed = True                                                         │
│   57 │   │   │   │   break                                                                   │
│                                                                                              │
│ /home1/ai20resch16001/clup/cli/commands/clup.py:143 in run                                   │
│                                                                                              │
│   140 │   │   │   criterion_kw["reduction"] = criterion_kw.get("reduction", "batchmean")     │
│   141 │   │   │   criterion = torch.nn.KLDivLoss(**criterion_kw)                             │
│   142 │   │                                                                                  │
│ ❱ 143 │   │   system = CluPSystem(                                                           │
│   144 │   │   │   model,                                                                     │
│   145 │   │   │   labels,                                                                    │
│   146 │   │   │   labeled_samples=subset_idxs,                                               │
│                                                                                              │
│ /home1/ai20resch16001/clup/src/systems/clup.py:30 in __init__                                │
│                                                                                              │
│    27 │   ) -> None:                                                                         │
│    28 │   │   print("before init in systems.clup")                                           │
│    29 │   │   print("model: ", model)                                                        │
│ ❱  30 │   │   super().__init__(model, *args, pseudo_every=1, **kwargs)                       │
│    31 │   │   self.labels = labels                                                           │
│    32 │   │   self.labeled_samples: List[int] = labeled_samples or []                        │
│    33 │   │   self.train_criterion = self.criterion                                          │
│                                                                                              │
│ /home1/ai20resch16001/clup/src/systems/mixins/pseudo_labelling.py:20 in __init__             │
│                                                                                              │
│    17 │   """Implementation of a mixin for pseudo labelling."""                              │
│    18 │                                                                                      │
│    19 │   def __init__(self, *args, **kwargs) -> None:                                       │
│ ❱  20 │   │   super().__init__(*args, **kwargs)                                              │
│    21 │   │   self.pseudo_labels: Tensor                                                     │
│    22 │   │   self.pseudo_labels_confidence: Tensor                                          │
│    23 │   │   self.pseudo_every = kwargs.get("pseudo_every", 10)                             │
│                                                                                              │
│ /home1/ai20resch16001/clup/src/systems/mixins/mixup.py:13 in __init__                        │
│                                                                                              │
│   10 │   """Implementation of a mixin for mixup."""                                          │
│   11 │                                                                                       │
│   12 │   def __init__(self, *args, **kwargs) -> None:                                        │
│ ❱ 13 │   │   super().__init__(*args, **kwargs)                                               │
│   14 │   │   self.mixup_alpha = kwargs.get("mixup_alpha", -1)                                │
│   15 │   │   self.mixup_last_lambda = -1                                                     │
│   16                                                                                         │
│                                                                                              │
│ /home1/ai20resch16001/clup/src/systems/mixins/base.py:15 in __init__                         │
│                                                                                              │
│   12 │   """Implementation of an abstract mixin class."""                                    │
│   13 │                                                                                       │
│   14 │   def __init__(self, *args, **kwargs) -> None:                                        │
│ ❱ 15 │   │   super().__init__(*args, **kwargs)                                               │
│   16 │   │                                                                                   │
│   17 │   │   if not hasattr(self, "metrics"):                                                │
│   18 │   │   │   self.metrics: Dict[str, MetricCollection]                                   │
│                                                                                              │
│ /home1/ai20resch16001/clup/src/systems/classification.py:38 in __init__                      │
│                                                                                              │
│    35 │   │   :param lr: learning rate of the system                                         │
│    36 │   │   :param lr_scheduler: learning rate policy for training                         │
│    37 │   │   """                                                                            │
│ ❱  38 │   │   super().__init__(model, *args, **kwargs)                                       │
│    39 │   │                                                                                  │
│    40 │   │   self.criterion: Module = torch.nn.CrossEntropyLoss()                           │
│    41 │   │   if criterion:                                                                  │
│                                                                                              │
│ /home1/ai20resch16001/clup/src/systems/base.py:32 in __init__                                │
│                                                                                              │
│    29 │   │   self.kwargs = kwargs                                                           │
│    30 │   │   self.save_hyperparameters(kwargs)                                              │
│    31 │   │                                                                                  │
│ ❱  32 │   │   print("model.encoder in systems.base#######", model.encoder)                   │
│    33 │   │                                                                                  │
│    34 │   │   self.encoder = model.encoder                                                   │
│    35 │   │   self.classifier = model.classifier                                             │
╰──────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'NoneType' object has no attribute 'encoder'

**Comment by altndrr** [comment]
Can you share the env? What command did you execute to get this output?

**Comment by KhinChoWin-AI** [comment]
The env is as follows:
name: base
channels:
  - defaults
dependencies:
  - conda
  - python=3.9
  - conda-env
  - conda-build
  - conda-verify
  - _ipyw_jlab_nb_ext_conf
  - anaconda==2022.05=py39_0
  - anaconda-navigator
  - navigator-updater
  - pillow
  - cudatoolkit=11.2
  - cudnn=8.1.0
prefix: /home1/ai20resch16001/anaconda3


The command is python -m cli clup \
    ./models/target/pretrained/swav-1000ep-FER2013-5000prototypes.ckpt \
    ./models/source/resnet18_rafdb.pt \
    fer2013 \
    --epochs 50 \
    --lr 0.1 \
    --batch_size 32 \
    --optimizer sgd \
    --optimizer_kw "{\"weight_decay\": 0.0005, \"momentum\": 0.9, \"nesterov\": true}" \
    --lr_scheduler cosine_annealing \
    --lr_scheduler_kw "{\"T_max\": 50}" \
    --mixup_alpha 0.2 \
    --labeled_samples_mode purity \
    --labeled_samples_mode_kw "{\"num_prototypes\": 250, \"drop_outliers\": false, \"iter_clusters\": 1}" \
    --labeled_samples_size 0.3

**Comment by altndrr** [comment]
From this I am not sure I can check the list of packages installed. Did you follow the instructions to install? Did you then also source the venv? Just making sure before moving on with other possible causes.

**Comment by KhinChoWin-AI** [comment]
Yes, I also followed the same stepup that you mentioned using .venv.

**Comment by altndrr** [comment]
Have you tried other commands? Is this the only one that does not work? Can you confirm the two args pointing to model weights link to existing files?
